Nil pointer deref in `pkg/datapath/loader/compile.go`

@lmb Looks like https://github.com/cilium/cilium/commit/d2ee07e80526ccb89a2da17d67325d279905a461 introduced a nil dereference, at first glance `compileCmd.ProcessState` may be nil. (`Cmd.Wait()` needs to be called first)
```
level=debug msg="Launching compiler" args="[-I/var/run/cilium/state/globals -I/var/run/cilium/state/templates/48d2511cca9b589bf5c6ec7d63c3775039557a2c3a6f3a95a3505558361bade3 -I/var/lib/cilium/bpf -I/var/lib/cilium/bpf/include -g -O2 --target=bpf -std=gnu89 -nostdinc -D__NR_CPUS__=2 -Wall -Wextra -Werror -Wshadow -Wno-address-of-packed-member -Wno-unknown-warning-option -Wno-gnu-variable-sized-type-not-at-end -Wdeclaration-after-statement -Wimplicit-int-conversion -Wenum-conversion -mcpu=v3 -c /var/lib/cilium/bpf/bpf_lxc.c -o -]" subsys=datapath-loader target=clang
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1f3a828]

goroutine 1645 [running]:
os.(*ProcessState).sysUsage(...)
        /usr/local/go/src/os/exec_posix.go:95
os.(*ProcessState).SysUsage(...)
        /usr/local/go/src/os/exec.go:179
github.com/cilium/cilium/pkg/datapath/loader.compile({0x38dcae0, 0x4002e9e770}, 0x5993660, 0x4002e5db90)
        /go/src/github.com/cilium/cilium/pkg/datapath/loader/compile.go:203 +0x978
github.com/cilium/cilium/pkg/datapath/loader.compileDatapath({0x38dcae0?, 0x4002e9e770}, 0x400168e650?, 0x0, 0x4000b43098?)
        /go/src/github.com/cilium/cilium/pkg/datapath/loader/compile.go:275 +0x520
github.com/cilium/cilium/pkg/datapath/loader.compileTemplate(...)
        /go/src/github.com/cilium/cilium/pkg/datapath/loader/compile.go:319
github.com/cilium/cilium/pkg/datapath/loader.(*objectCache).build(0x40021e14f0, {0x38dcae0, 0x4002e9e770}, 0x4000b43098, {0x4002e7f500, 0x40})
        /go/src/github.com/cilium/cilium/pkg/datapath/loader/cache.go:255 +0x480
github.com/cilium/cilium/pkg/datapath/loader.(*objectCache).fetchOrCompile.func2()
        /go/src/github.com/cilium/cilium/pkg/datapath/loader/cache.go:306 +0xf4
github.com/cilium/cilium/pkg/serializer.(*FunctionQueue).run(0x4000bbf300)
        /go/src/github.com/cilium/cilium/pkg/serializer/func_queue.go:29 +0x6c
created by github.com/cilium/cilium/pkg/serializer.NewFunctionQueue in goroutine 1463
        /go/src/github.com/cilium/cilium/pkg/serializer/func_queue.go:19 +0xe4
```

--- a/pkg/datapath/loader/compile.go
+++ b/pkg/datapath/loader/compile.go
@@ -202,20 +202,12 @@ func compile(ctx context.Context, prog *progInfo, dir *directoryInfo) (string, e
 	var compilerStderr bytes.Buffer
 	compileCmd.Stderr = &compilerStderr
 
-	err = compileCmd.Run()
-
-	var maxRSS int64
-	if usage, ok := compileCmd.ProcessState.SysUsage().(*syscall.Rusage); ok {
-		maxRSS = usage.Maxrss
-	}
-
-	if err != nil {
+	if err := compileCmd.Run(); err != nil {
 		err = fmt.Errorf("Failed to compile %s: %w", prog.Output, err)
 
 		if !errors.Is(err, context.Canceled) {
 			log.WithFields(logrus.Fields{
 				"compiler-pid": pidFromProcess(compileCmd.Process),
-				"max-rss":      maxRSS,
 			}).Error(err)
 		}
 
@@ -227,11 +219,14 @@ func compile(ctx context.Context, prog *progInfo, dir *directoryInfo) (string, e
 		return "", err
 	}
 
-	if maxRSS > 0 {
+	// Cmd.ProcessState is populated by Cmd.Wait(). Cmd.Run() bails out if
+	// Cmd.Start() fails, which will leave Cmd.ProcessState nil. Only log peak
+	// RSS if the compilation succeeded, which will be the majority of cases.
+	if usage, ok := compileCmd.ProcessState.SysUsage().(*syscall.Rusage); ok {
 		log.WithFields(logrus.Fields{
 			"compiler-pid": compileCmd.Process.Pid,
 			"output":       output.Name(),
-		}).Debugf("Compilation had peak RSS of %d bytes", maxRSS)
+		}).Debugf("Compilation had peak RSS of %d bytes", usage.Maxrss)
 	}
 
 	return output.Name(), nil

